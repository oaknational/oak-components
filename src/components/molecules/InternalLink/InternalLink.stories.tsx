import React from "react";
import { Meta, StoryObj } from "@storybook/react";

import { InternalLink } from "./InternalLink";

const meta: Meta<typeof InternalLink> = {
  component: InternalLink,
  tags: ["autodocs"],
  title: "components/molecules/InternalLink",
  argTypes: {
    children: { type: "string" },
    element: { type: "string" },
  },
  parameters: {
    controls: {
      exclude: ["ref"],
    },
  },
  args: {
    children: "Link",
    color: "text-link-active",
    hoverColor: "text-link-hover",
    activeColor: "text-link-pressed",
    disabledColor: "text-disabled",
    visitedColor: "text-link-visited",
  },
  render: (args) => <InternalLink {...args} />,
};
export default meta;

type Story = StoryObj<typeof InternalLink>;

export const Default: Story = {
  args: {
    href: `#${Math.random()}`,
  },
};

export const AsAButton: Story = {
  args: {
    element: "button",
  },
};

export const WithIcon: Story = {
  args: {
    href: `#${Math.random()}`,
    iconName: "chevron-left",
  },
};

export const WithTrailingIcon: Story = {
  args: {
    href: `#${Math.random()}`,
    iconName: "chevron-right",
    isTrailingIcon: true,
  },
};

export const Disabled: Story = {
  args: {
    element: "button",
    disabled: true,
    iconName: "chevron-left",
  },
};

export const Loading: Story = {
  args: {
    element: "button",
    children: "Loading...",
    isLoading: true,
  },
};
