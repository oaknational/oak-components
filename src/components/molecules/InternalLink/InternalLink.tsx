import React, { ElementType, forwardRef } from "react";
import styled from "styled-components";

import { OakLoadingSpinner } from "@/components/molecules/OakLoadingSpinner";
import {
  PolymorphicPropsWithRef,
  PolymorphicRef,
} from "@/components/polymorphic";
import { parseColor } from "@/styles/helpers/parseColor";
import { parseDropShadow } from "@/styles/helpers/parseDropShadow";
import { parseBorderRadius } from "@/styles/helpers/parseBorderRadius";
import { parseSpacing } from "@/styles/helpers/parseSpacing";
import { OakFlex, OakIcon, OakIconProps, OakSpan } from "@/components/atoms";
import { parseColorFilter } from "@/styles/helpers/parseColorFilter";
import { OakAllSpacingToken, OakCombinedColorToken } from "@/styles";

const StyledOakIcon = styled(OakIcon)``;

const StyledLink = styled.a<{
  $color: OakCombinedColorToken;
  $visitedColor: OakCombinedColorToken;
  $hoverColor: OakCombinedColorToken;
  $activeColor: OakCombinedColorToken;
  $disabledColor: OakCombinedColorToken;
}>`
  display: inline-flex;
  align-items: center;
  gap: ${parseSpacing("space-between-sssx")};
  outline: none;
  border-radius: ${parseBorderRadius("border-radius-m")};
  padding: ${parseSpacing("inner-padding-ssx")};
  margin: -${parseSpacing("inner-padding-ssx")};
  appearance: none;
  font: inherit;
  background: none;
  border: none;
  text-decoration: underline;
  cursor: pointer;
  color: ${(props) => parseColor(props.$color)};

  ${StyledOakIcon} {
    filter: ${(props) => parseColorFilter(props.$color)};
  }

  &:focus-visible {
    box-shadow: ${parseDropShadow("drop-shadow-centered-lemon")},
      ${parseDropShadow("drop-shadow-centered-grey")};
  }

  &:visited {
    color: ${(props) => parseColor(props.$visitedColor)};

    ${StyledOakIcon} {
      filter: ${(props) => parseColorFilter(props.$visitedColor)};
    }
  }

  @media (hover: hover) {
    &:hover,
    &:visited:hover {
      color: ${(props) => parseColor(props.$hoverColor)};
      ${StyledOakIcon} {
        filter: ${(props) => parseColorFilter(props.$hoverColor)};
      }
    }
  }

  &:active {
    color: ${(props) => parseColor(props.$activeColor)};

    ${StyledOakIcon} {
      filter: ${(props) => parseColorFilter(props.$activeColor)};
    }
  }

  &[disabled] {
    cursor: not-allowed;
    color: ${(props) => parseColor(props.$disabledColor)};

    ${StyledOakIcon} {
      filter: ${(props) => parseColorFilter(props.$disabledColor)};
    }
  }
`;

export type InternalLinkProps = {
  /**
   * The icon to display before or after the children.
   */
  iconName?: OakIconProps["iconName"];
  /**
   * Whether the icon should be displayed after the children.
   */
  isTrailingIcon?: boolean;
  /**
   * Displays a loading spinner in place of the icon.
   */
  isLoading?: boolean;
  color: OakCombinedColorToken;
  visitedColor: OakCombinedColorToken;
  hoverColor: OakCombinedColorToken;
  activeColor: OakCombinedColorToken;
  disabledColor: OakCombinedColorToken;
  iconWidth?: OakAllSpacingToken;
  iconHeight?: OakAllSpacingToken;
};

type InternalLinkComponent = <C extends React.ElementType = "a">(
  props: PolymorphicPropsWithRef<C> & InternalLinkProps,
) => React.ReactNode;

/**
 * A link with an optional icon and loading state.
 *
 * Defaulting to a `HTMLAnchorElement` this component is polymorphic and can be rendered as a button or any other element.
 */
export const InternalLink: InternalLinkComponent = forwardRef(
  <C extends ElementType = "a">(
    props: PolymorphicPropsWithRef<C> & InternalLinkProps,
    ref: PolymorphicRef<C>,
  ) => {
    const {
      element = "a",
      iconName,
      isTrailingIcon,
      children,
      isLoading,
      color,
      disabledColor,
      visitedColor,
      hoverColor,
      activeColor,
      iconWidth = "all-spacing-6",
      iconHeight = "all-spacing-6",
      ...rest
    } = props;

    const icon = (() => {
      switch (true) {
        case isLoading:
          return (
            <OakFlex
              $width="all-spacing-6"
              $height="all-spacing-6"
              $alignItems="center"
              $justifyContent="center"
            >
              <OakLoadingSpinner
                $width="all-spacing-4"
                $color="icon-inverted"
              />
            </OakFlex>
          );
        case !!iconName:
          return (
            <StyledOakIcon
              iconName={iconName}
              iconWidth={iconWidth}
              iconHeight={iconHeight}
            />
          );
        default:
          return null;
      }
    })();
    const disabled = (() => {
      switch (true) {
        case "disabled" in rest:
          return rest.disabled;
        case isLoading:
          return true;
        default:
          return undefined;
      }
    })();

    return (
      <StyledLink
        as={element}
        ref={ref}
        disabled={disabled}
        $color={color}
        $disabledColor={disabledColor}
        $visitedColor={visitedColor}
        $hoverColor={hoverColor}
        $activeColor={activeColor}
        {...rest}
      >
        {!isTrailingIcon && icon}
        <OakSpan>{children}</OakSpan>
        {isTrailingIcon && icon}
      </StyledLink>
    );
  },
);
