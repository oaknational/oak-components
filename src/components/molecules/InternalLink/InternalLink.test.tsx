import React from "react";
import "@testing-library/jest-dom";

import { InternalLink } from "./InternalLink";

import renderWithTheme from "@/test-helpers/renderWithTheme";

describe("InternalLink", () => {
  it("matches snapshot", () => {
    const { container } = renderWithTheme(
      <InternalLink
        color="text-link-active"
        hoverColor="text-link-hover"
        activeColor="text-link-pressed"
        disabledColor="text-disabled"
        visitedColor="text-link-visited"
      >
        Content goes here
      </InternalLink>,
    );

    expect(container).toMatchSnapshot();
  });
});
