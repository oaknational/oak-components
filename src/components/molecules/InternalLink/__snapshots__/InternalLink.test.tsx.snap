// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InternalLink matches snapshot 1`] = `
.c1 {
  font-family: --var(google-font),Lexend,sans-serif;
}

.c0 {
  display: inline;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  gap: 0.25rem;
  outline: none;
  border-radius: 0.375rem;
  padding: 0.25rem;
  margin: -0.25rem;
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
  font: inherit;
  background: none;
  border: none;
  -webkit-text-decoration: underline;
  text-decoration: underline;
  cursor: pointer;
  color: #0d24c4;
}

.c0:focus-visible {
  box-shadow: 0 0 0 0.125rem rgba(255,229,85,100%), 0 0 0 0.3rem rgba(87,87,87,100%);
}

.c0:visited {
  color: #081676;
}

.c0:active {
  color: #081676;
}

.c0[disabled] {
  cursor: not-allowed;
  color: #808080;
}

@media (hover:hover) {
  .c0:hover,
  .c0:visited:hover {
    color: #0a1d9d;
  }
}

<div>
  <a
    class="c0"
  >
    <span
      class="c1"
    >
      Content goes here
    </span>
  </a>
</div>
`;
